File: cpp/leetcode/python/editDistance.py
#!/usr/bin/python 
# -*- coding: utf-8 -*- 
# Given two words word1 and word2, find the minimum number of steps required to convert word1 to
# word2. (each operation is counted as 1 step.)
# You have the following 3 operations permied on a word:
# • Insert a character
# • Delete a character
# • Replace a characte

def minDistance(word1, word2):
	'''由于我们在DP过程中仅仅是对上一行的数据操作获得下一行的数据，
	我们完全可以考虑将不重用的数据空间提前释放或者申请适量的数据空间
	而不是申请一个二维数组空间进行存储数据
	'''
	if(len(word1) < len(word2)):
		return minDistance(word2, word1)
	l = len(word1)
	
	# 初始化动态规划数组空间，相当于对于word2 == 0的时候，对于word1不同位置
	# 的时候对应的编辑距离， 因此，结果应该存储在f[l]的位置
	f = [i for i in range(l+1)]
	
	# DP获取最小编辑距离
	for i in range(1, l+1):
		upper_left = f[0] # save the f[i-1][j-1]
		f[0] = i
		for j in range(1, l+1):
			upper = f[j]
			if word1[i-1] == word2[j-1]:
				f[j] = upper_left
			else:
				f[j] = 1 + min(upper_left, min(f[j], f[j-1]))
			upper_left = upper
	return f[l]

def test_minDistance():
	word1 = "hjllo"
	word2 = "hello"
	assert minDistance(word1, word2) == 1

if __name__ == '__main__':
	test_minDistance()
			

File: cpp/leetcode/book/dp/edit_distance.cpp
/**********************************************************************************************
    Given two words word1 and word2, find the minimum number of steps required to convert word1 to
    word2. (each operation is counted as 1 step.)
    You have the following 3 operations permitted on a word:
    • Insert a character
    • Delete a character
    • Replace a character
/**********************************************************************************************/
#include <iostream>
#include <string>
#include <cassert>
using namespace std;
class Solution {
    public: int minDistance(const string & word1, const string & word2) {
        const size_t n = word1.size();
        const size_t m = word2.size();
        // 长度为 n 的字符串，有 n+1 个隔板
        int f[n + 1][m + 1];
        for (size_t i = 0; i <= n; i++) f[i][0] = i;
        for (size_t j = 0; j <= m; j++) f[0][j] = j;

        for (size_t i = 1; i <= n; i++) // word1 iter
        {
            for (size_t j = 1; j <= m; j++) // word2 iter
            {
                if (word1[i - 1] == word2[j - 1]) 
                	f[i][j] = f[i - 1][j - 1];
                else 
                {
                    int mn = min(f[i - 1][j], f[i][j - 1]);
                    f[i][j] = 1 + min(f[i - 1][j - 1], mn);
                }
            }
        }
        return f[n][m];
    }
};
int main(int argc, char const *argv[])
{
	Solution solve;
	string a = "hello";
	string b = "2hello2";
	assert(solve.minDistance(a, b) == 2);
	assert(solve.minDistance("helllo","hel2o") == 2);
	assert(solve.minDistance("hello","") == 5);
	assert(solve.minDistance("hehe","hhehhe") == 2);
	std::cout << "TEST PASSED"<<std::endl;
	return 0;
}